nemo/collections/llm/inference/base.py

```diff
@@ -62,19 +64,23 @@ def _setup_trainer_and_restore_model(path: Path, trainer: nl.Trainer, model: pl.
 def setup_model_and_tokenizer(
     path: Path,
     trainer: Optional[nl.Trainer] = None,
+    tensor_parallel_size: int = -1,
     params_dtype: torch.dtype = torch.bfloat16,
     inference_batch_times_seqlen_threshold: int = 1000,
 ) -> tuple[MCoreGPTModel, MCoreTokenizerWrappper]:
-    model: io.TrainerContext = io.load_context(path=path, subpath="model")
-    trainer = trainer or io.load_context(path=path, subpath="trainer")
+    model: io.TrainerContext = io.load_context(path=ckpt_to_context_subdir(path), subpath="model")
+    trainer = trainer or io.load_context(path=ckpt_to_context_subdir(path), subpath="trainer")
+    if tensor_parallel_size > 0:
+        trainer.strategy.tensor_model_parallel_size = tensor_parallel_size
+        trainer.devices = tensor_parallel_size
+        trainer.strategy.parallel_devices = [torch.device(f"cuda:{i}") for i in range(tensor_parallel_size)]
```

Comment:
Do we need this? I don't think we need to set this explicitly.

Reply:
If we don't set explicitly, the tp will be inherit from load_context's trainer, which will be SFT training setting, which is greater than what we need when just doing llm.generate. Do we want user to be able to set it?

Reply:
Oh I meant just for `trainer.strategy.parallel_devices`

Reply:
If I don't change strategy.parallel_devices, both strategy.parallel_devices and launcher.num_processes will be the SFT training setting--it is init/built at load_context time. When using torchrun, the launcher will launch SFT tp setting and complain about world size mismatch. Do you think it's ok or too hacky and maybe directly modify the saved SFT context?